File: backend/app/routers/system_actions.py
"""SuperAdmin remote-control actions.

Delivers via the mobile app the ability to restart containers, view logs,
trigger backups, etc. — without SSH.

Security model:
- All endpoints require SUPER_ADMIN (NOT ADMIN). Highest blast radius, tightest gate.
- Every action audit-logged via UserActivityLog.
- Container ops are whitelist-restricted (admin-backend / admin-frontend, or prod equivalents).
- Rate-limited via slowapi.

Mounting required (docker-compose.admin.yml admin-backend service):
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - /var/lib/ssmspl-host-actions:/var/lib/ssmspl-host-actions   # optional, for v2.2 host actions
"""
from __future__ import annotations

import logging
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, Field
from sqlalchemy import update
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.rbac import UserRole
from app.database import get_db
from app.dependencies import require_roles
from app.middleware.rate_limit import limiter
from app.models.system_health_event import SystemHealthEvent
from app.models.user import User
from app.services import host_action_service, push_service, system_actions_service
from app.services.activity_log_service import ActivityAction, log_activity

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/system-health/actions", tags=["System Health · Actions"])

_super_admin_only = require_roles(UserRole.SUPER_ADMIN)


# ─── schemas ─────────────────────────────────────────────────────────


class ContainerActionBody(BaseModel):
    name: str = Field(..., max_length=80)


class ActionResult(BaseModel):
    ok: bool
    detail: dict | None = None
    error: str | None = None


# ─── helpers ─────────────────────────────────────────────────────────


def _check_container(name: str) -> None:
    allowed = system_actions_service.allowed_containers()
    if name not in allowed:
        raise HTTPException(
            status_code=400,
            detail=f"container {name!r} not allowed; permitted: {sorted(allowed)}",
        )


# ─── docker-socket actions ───────────────────────────────────────────


@router.post("/restart-container", response_model=ActionResult)
async def restart_container(
    body: ContainerActionBody,
    current_user: Annotated[User, Depends(_super_admin_only)],
):
    _check_container(body.name)
    try:
        result = system_actions_service.restart_container(body.name)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:  # noqa: BLE001
        logger.exception("restart_container failed")
        return ActionResult(ok=False, error=str(e))
    await log_activity(
        getattr(current_user, "active_session_id", None),
        current_user.id,
        ActivityAction.SYSTEM_RESTART_CONTAINER,
        {"container": body.name, "duration_s": result.get("duration_s")},
    )
    return ActionResult(ok=True, detail=result)


@router.post("/prune-images", response_model=ActionResult)
async def prune_images(
    current_user: Annotated[User, Depends(_super_admin_only)],
):
    result = system_actions_service.prune_docker_images()
    await log_activity(
        getattr(current_user, "active_session_id", None),
        current_user.id,
        ActivityAction.SYSTEM_PRUNE_IMAGES,
        result,
    )
    return ActionResult(ok=result.get("ok", False), detail=result, error=result.get("error"))


# ─── filesystem actions (no docker socket needed) ────────────────────


@router.post("/trigger-backup", response_model=ActionResult)
async def trigger_backup(
    current_user: Annotated[User, Depends(_super_admin_only)],
):
    result = system_actions_service.trigger_backup()
    await log_activity(
        getattr(current_user, "active_session_id", None),
        current_user.id,
        ActivityAction.SYSTEM_TRIGGER_BACKUP,
        result,
    )
    return ActionResult(ok=result.get("ok", False), detail=result, error=result.get("error"))


@router.post("/force-sync", response_model=ActionResult)
async def force_sync(
    current_user: Annotated[User, Depends(_super_admin_only)],
):
    result = system_actions_service.force_sync()
    await log_activity(
        getattr(current_user, "active_session_id", None),
        current_user.id,
        ActivityAction.SYSTEM_FORCE_SYNC,
        result,
    )
    return ActionResult(ok=result.get("ok", False), detail=result, error=result.get("error"))


# ─── DB-side actions ─────────────────────────────────────────────────


@router.post("/test-push", response_model=ActionResult)
async def test_push(
    current_user: Annotated[User, Depends(_super_admin_only)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """Fire a test push to all registered devices. Useful to verify the chain after install."""
    result = await push_service.send_push(
        db,
        title="SSMSPL — test push",
        body="If you see this, your SuperAdmin app + backend chain is healthy.",
        data={"kind": "test"},
    )
    await log_activity(
        getattr(current_user, "active_session_id", None),
        current_user.id,
        ActivityAction.SYSTEM_TEST_PUSH,
        result,
    )
    return ActionResult(ok=result["sent"] > 0, detail=result)


class AckBody(BaseModel):
    note: str | None = Field(None, max_length=500)


@router.post("/events/{event_id}/ack", response_model=ActionResult)
async def ack_event(
    event_id: int,
    current_user: Annotated[User, Depends(_super_admin_only)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """Mark an alert as acknowledged (read). Doesn't delete; just hides from unacked queries."""
    from sqlalchemy import func

    r = await db.execute(
        update(SystemHealthEvent)
        .where(SystemHealthEvent.id == event_id)
        .where(SystemHealthEvent.acked_at.is_(None))
        .values(acked_at=func.now(), acked_by=current_user.id)
        .returning(SystemHealthEvent.id)
    )
    if r.scalar() is None:
        raise HTTPException(status_code=404, detail="Event not found or already acked")
    await db.commit()
    await log_activity(
        getattr(current_user, "active_session_id", None),
        current_user.id,
        ActivityAction.SYSTEM_ACK_EVENT,
        {"event_id": event_id},
    )
    return ActionResult(ok=True, detail={"event_id": event_id})


# ─── container introspection ─────────────────────────────────────────


@router.get("/containers")
async def list_containers(_user: Annotated[User, Depends(_super_admin_only)]):
    """List the whitelisted containers + their health (via docker socket)."""
    out = []
    for name in sorted(system_actions_service.allowed_containers()):
        try:
            out.append(system_actions_service.get_container_inspect(name))
        except Exception as e:  # noqa: BLE001
            out.append({"name": name, "error": str(e)})
    return out


@router.get("/containers/{name}/logs")
async def container_logs(
    name: str,
    _user: Annotated[User, Depends(_super_admin_only)],
    lines: int = Query(100, ge=1, le=2000),
):
    _check_container(name)
    try:
        log_lines = system_actions_service.container_logs(name, lines=lines)
    except Exception as e:  # noqa: BLE001
        raise HTTPException(status_code=500, detail=str(e))
    return {"name": name, "lines": log_lines, "count": len(log_lines)}


@router.get("/containers/{name}/stats")
async def container_stats(name: str, _user: Annotated[User, Depends(_super_admin_only)]):
    _check_container(name)
    try:
        return system_actions_service.container_stats(name)
    except Exception as e:  # noqa: BLE001
        raise HTTPException(status_code=500, detail=str(e))


# ─── host-daemon actions (only enabled when daemon + queue volume present) ─


@router.get("/host-daemon-status", response_model=ActionResult)
async def host_daemon_status(_user: Annotated[User, Depends(_super_admin_only)]):
    """Tells the app whether host-action endpoints will work."""
    return ActionResult(
        ok=host_action_service.is_queue_mounted(),
        detail={
            "queue_mounted": host_action_service.is_queue_mounted(),
            "queue_root": str(host_action_service.QUEUE_ROOT),
            "allowed_actions": sorted(host_action_service.ALLOWED_ACTIONS),
        },
    )


class HostActionBody(BaseModel):
    action: str = Field(..., max_length=80)
    params: dict = Field(default_factory=dict)
    timeout_s: float = Field(30.0, ge=2, le=120)


@router.post("/host", response_model=ActionResult)
async def submit_host_action(
    body: HostActionBody,
    current_user: Annotated[User, Depends(_super_admin_only)],
):
    """Submit a host-side action via queue daemon (run_iptables_fix, force_recreate_admin_backend, etc.)."""
    if body.action not in host_action_service.ALLOWED_ACTIONS:
        raise HTTPException(
            status_code=400,
            detail=f"action {body.action!r} not whitelisted; allowed: {sorted(host_action_service.ALLOWED_ACTIONS)}",
        )
    result = await host_action_service.submit_and_wait(body.action, body.params, timeout_s=body.timeout_s)
    action_const = {
        "kill_pid": ActivityAction.SYSTEM_RESTART_CONTAINER,
        "force_recreate_admin_backend": ActivityAction.SYSTEM_RESTART_CONTAINER,
        "run_iptables_fix": ActivityAction.SYSTEM_IPTABLES_FIX,
        "run_health_check": ActivityAction.SYSTEM_RUN_HEALTH_CHECK,
        "cleanup_logs": ActivityAction.SYSTEM_PRUNE_IMAGES,
    }.get(body.action, ActivityAction.SYSTEM_HOST_ACTION)
    await log_activity(
        getattr(current_user, "active_session_id", None),
        current_user.id,
        action_const,
        {"action": body.action, "params": body.params, "request_id": result.get("request_id")},
    )
    return ActionResult(ok=result.get("ok", False), detail=result, error=result.get("error"))

File: mobile/src/api/systemActions.ts
import { getClient } from './client';

const ACTIONS_BASE = '/api/system-health/actions';

export type ActionResult = {
  ok: boolean;
  detail?: Record<string, unknown> | null;
  error?: string | null;
};

export type ContainerInspect = {
  name: string;
  id?: string;
  status?: string;
  health?: string | null;
  restart_count?: number;
  started_at?: string;
  image?: string;
  error?: string;
};

export type ContainerStats = {
  name: string;
  cpu_pct: number;
  mem_used_mb: number;
  mem_limit_mb: number;
};

export type HostDaemonStatus = {
  ok: boolean;
  detail?: {
    queue_mounted: boolean;
    queue_root: string;
    allowed_actions: string[];
  };
};

// ─── container actions ──────────────────────────────────────────────

export async function listContainers(): Promise<ContainerInspect[]> {
  const c = await getClient();
  const r = await c.get(`${ACTIONS_BASE}/containers`);
  return r.data;
}

export async function tailContainerLogs(name: string, lines = 200): Promise<{ name: string; lines: string[]; count: number }> {
  const c = await getClient();
  const r = await c.get(`${ACTIONS_BASE}/containers/${encodeURIComponent(name)}/logs`, { params: { lines } });
  return r.data;
}

export async function getContainerStats(name: string): Promise<ContainerStats> {
  const c = await getClient();
  const r = await c.get(`${ACTIONS_BASE}/containers/${encodeURIComponent(name)}/stats`);
  return r.data;
}

export async function restartContainer(name: string): Promise<ActionResult> {
  const c = await getClient();
  const r = await c.post(`${ACTIONS_BASE}/restart-container`, { name });
  return r.data;
}

export async function pruneImages(): Promise<ActionResult> {
  const c = await getClient();
  const r = await c.post(`${ACTIONS_BASE}/prune-images`);
  return r.data;
}

// ─── filesystem actions ─────────────────────────────────────────────

export async function triggerBackup(): Promise<ActionResult> {
  const c = await getClient();
  const r = await c.post(`${ACTIONS_BASE}/trigger-backup`);
  return r.data;
}

export async function forceSync(): Promise<ActionResult> {
  const c = await getClient();
  const r = await c.post(`${ACTIONS_BASE}/force-sync`);
  return r.data;
}

// ─── DB actions ─────────────────────────────────────────────────────

export async function testPush(): Promise<ActionResult> {
  const c = await getClient();
  const r = await c.post(`${ACTIONS_BASE}/test-push`);
  return r.data;
}

export async function ackEvent(eventId: number): Promise<ActionResult> {
  const c = await getClient();
  const r = await c.post(`${ACTIONS_BASE}/events/${eventId}/ack`);
  return r.data;
}

export async function ackAllEvents(): Promise<ActionResult> {
  const c = await getClient();
  const r = await c.post(`${ACTIONS_BASE}/events/ack-all`);
  return r.data;
}

// ─── host-daemon actions ────────────────────────────────────────────

export async function getHostDaemonStatus(): Promise<HostDaemonStatus> {
  const c = await getClient();
  const r = await c.get(`${ACTIONS_BASE}/host-daemon-status`);
  return r.data;
}

export async function submitHostAction(
  action: string,
  params: Record<string, unknown> = {},
  timeoutS = 30,
): Promise<ActionResult> {
  const c = await getClient();
  const r = await c.post(`${ACTIONS_BASE}/host`, { action, params, timeout_s: timeoutS });
  return r.data;
}

// ─── version + releases + rollback ─────────────────────────────────

export type VersionInfo = {
  git_sha: string;
  build_ts: string;
  alembic_head: string;
  image_tag: string;
};

export type ReleaseEntry = {
  image_tag: string;
  git_sha: string;
  build_ts: string;
  alembic_head: string;
  deployed_at?: string;
  deployed_by?: string;
  host?: string;
  image_present: boolean;
  is_current: boolean;
};

export type ReleasesResponse = {
  current: VersionInfo;
  releases: ReleaseEntry[];
};

export async function getCurrentVersion(): Promise<VersionInfo> {
  const c = await getClient();
  const r = await c.get(`${ACTIONS_BASE}/version`);
  return r.data;
}

export async function listReleases(limit = 20): Promise<ReleasesResponse> {
  const c = await getClient();
  const r = await c.get(`${ACTIONS_BASE}/releases`, { params: { limit } });
  return r.data;
}

export async function rollbackToRelease(
  imageTag: string,
  forceSchemaDrift = false,
): Promise<ActionResult> {
  const c = await getClient();
  const r = await c.post(`${ACTIONS_BASE}/rollback`, {
    image_tag: imageTag,
    force_schema_drift: forceSchemaDrift,
  });
  return r.data;
}

// ─── incident report ───────────────────────────────────────────────

export type IncidentReport = {
  generated_at: string;
  version: VersionInfo;
  snapshot: any;
  containers: any[];
  container_logs: Record<string, string[]>;
  events: Array<{
    id: number;
    server_name: string;
    severity: string;
    check_name: string;
    message: string;
    created_at: string | null;
    acked_at: string | null;
  }>;
  activity: Array<{
    id: string | null;
    action_type: string;
    user_id: string | null;
    created_at: string | null;
    metadata: Record<string, unknown> | null;
  }>;
};

export async function fetchIncidentReport(logLines = 200): Promise<IncidentReport> {
  const c = await getClient();
  const r = await c.get(`${ACTIONS_BASE}/incident-report`, { params: { log_lines: logLines } });
  return r.data;
}
